### DISCUSSION TITLE
Trouble with duppa i2c rotary encoder-Board on Pi Pico W

### DISCUSSION THREAD

**Discussion by JohnnyTank** [micropython/micropython#10389] [RP2040 / Pico]
I am trying to get a duppa I2C -rotary encoder board V2.1 running.
On a Pico W on a pimoroni-board explorer.
the i2c-scan gives an adress.
After two days of searching the Internet I have still no clue of what is going wrong. 
Help is very welcome.
the start_encoder method  is working too, but if i am trying to read data from the encoder i am getting:

Traceback (most recent call last):
  File "<stdin>", line 84, in <module>
  File "<stdin>", line 52, in main
  File "i2cEncoderLibV2.py", line 171, in updateStatus
  File "i2cEncoderLibV2.py", line 615, in readEncoder8
OSError: [Errno 5] EIO
 I am using this code:

```
from pimoroni_i2c import PimoroniI2C
from pimoroni import PICO_EXPLORER_I2C_PINS
from picographics import PicoGraphics, DISPLAY_PICO_EXPLORER, PEN_P8
import i2cEncoderLibV2
from time import sleep_ms
from pimoroni import Button
from machine import Pin



display = PicoGraphics(display=DISPLAY_PICO_EXPLORER, pen_type=PEN_P8)

WIDTH, HEIGHT = display.get_bounds()
button_a = Button(12)
button_b = Button(13)
button_x = Button(14)
button_y = Button(15)

def EncoderChange():
  print ('Changed: %d' % (encoder.readCounter32()))  
    
def start_encoder(bus, adr):
    encoder = i2cEncoderLibV2.i2cEncoderLibV2(bus, adr)
    encconfig = (i2cEncoderLibV2.INT_DATA |
                 i2cEncoderLibV2.WRAP_ENABLE |
                 i2cEncoderLibV2.DIRE_RIGHT |
                 i2cEncoderLibV2.IPUP_ENABLE |
                 i2cEncoderLibV2.RMOD_X1 |
                 i2cEncoderLibV2.STD_ENCODER)
    encoder.begin(encconfig)
    encoder.writeCounter(0)
    encoder.writeMax(35)
    encoder.writeMin(-20)
    encoder.writeStep(1)
    encoder.writeAntibouncingPeriod(8)
    encoder.writeDoublePushPeriod(50)
    return(encoder)

def end_encoder(encoder):
    encoder.reset()
    

def main(encoder, display):
    button_a = Button(12)
    button_b = Button(13)
    button_x = Button(14)
    button_y = Button(15)
    a_pressed = False
    while not(a_pressed):
        a_pressed = button_a.read()
        if (encoder.updateStatus()):
            print('Encoder changed')
        sleep_ms(100)
    

pen_white = display.create_pen(255, 255, 255)
BG = display.create_pen(40, 40, 40)
#display.clear()
display.set_pen(BG)
display.clear()
display.set_pen(pen_white)
#display.circle(100, 100, 10)
display.set_font('bitmap8')
display.text("Encoder Test", 0, 0, scale=3)
display.update()


i2c = PimoroniI2C(**PICO_EXPLORER_I2C_PINS)
sleep_ms(100)
adr_room=i2c.scan()
if (len(adr_room) > 0):
    enc_1 = adr_room[0]
    print(hex(enc_1))
else:
    enc_1 = -1
print (adr_room)

if (enc_1 > 0):
    encoder = start_encoder(i2c, enc_1)
    #encoder.onChange = EncoderChange
    sleep_ms(100)
    end_encoder(encoder)
    main(encoder, display)
else:
    print('Fehler! - Bitte Platine stromlos setzen')
print('Programm `Ende')

```

I tried to fit the Encoder Library to micropython:
The Error occures in this part of the code:
``# read the encoder 1 byte #     
    def readEncoder8(self, add):
        print('reading from: %d' %(self.i2cadd))
        print('Subadress: %d' %(add))
        data = self.i2cbus.readfrom_mem(self.i2cadd, add, 1)
        print(data)
        #value = struct.unpack(">b", bytearray(data))
        #return value[0]
        return data``

**Comment by robert-hh** [comment]
`OSError: [Errno 5] EIO` is an communication error. Typical reasons:
- wrong I2C address used.
- wrong pins for scl and sda
- insufficient Pull-up resistors

You said that scan retrurns the address of the encoder. Is the same address used by the driver software?
